Aplib.Core/Belief/SampledMemoryBelief.cs

```diff
@@ -0,0 +1,128 @@
+            UpdateMode updateMode,
+            int framesToRemember
+        )
+            : base(reference, getObservationFromReference, framesToRemember)
+        {
+            _sampleInterval = sampleInterval;
+            _updateMode = updateMode;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SampledMemoryBelief{TReference, TObservation}"/> class with an object reference,
+        /// a function to generate/update the observation using the object reference,
+        /// and a condition on when the observation should be updated.
+        /// This belief also stores a limited amount of previous observation samples in memory.
+        /// Optionally, the belief can always store the most recent observation, regardless of the sample rate.
+        /// </summary>
+        /// <param name="reference">The reference used to generate/update the observation.</param>
+        /// <param name="getObservationFromReference">A function that takes a reference and generates/updates a observation.</param>
+        /// <param name="sampleInterval">
+        /// The sample rate of the memory.
+        /// One observation memory (i.e., snapshot) is stored every <c>sampleInterval</c>-th cycle.
+        /// </param>
+        /// <param name="updateMode">Specifies how this sampled memory belief should be updated.</param>
+        /// <param name="framesToRemember">The number of frames to remember back.</param>
+        /// <param name="shouldUpdate">A function that sets a condition on when the observation should be updated.</param>
+        public SampledMemoryBelief
+        (
+            TReference reference,
+            Func<TReference, TObservation> getObservationFromReference,
+            int sampleInterval,
+            UpdateMode updateMode,
+            int framesToRemember,
+            Func<bool> shouldUpdate
+        )
+            : base(reference, getObservationFromReference, framesToRemember, shouldUpdate)
+        {
+            _sampleInterval = sampleInterval;
+            _updateMode = updateMode;
+        }
+
+        /// <summary>
+        /// Determines whether the memory should be sampled.
+        /// One observation memory (i.e., snapshot) is stored every <c>sampleInterval</c>-th cycle.
+        /// </summary>
+        /// <returns>Whether a memory sample should be stored in the current cycle.</returns>
+        private bool ShouldSampleMemory() => Counter++ == 0;
+
+        /// <summary>
+        /// Generates/updates the observation if applicable.
+        /// Also stores the previous observation in memory every <c>sampleInterval</c>-th cycle.
+        /// </summary>
+        public override void UpdateBelief()
+        {
+            if (ShouldSampleMemory())
+            {
+                // We use the implicit conversion to TObservation to store the observation.
+                _memorizedObservations.Put(this);
+
+                if (_shouldUpdate()) UpdateObservation();
```

Comment:
This is the implementation of `MemoryBelief.UpdateBelief()`. To avoid code-duplication, we should use `base.UpdateBelief()`.